text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,336 @@
+template.
+
+```ts
+assert.assertMatchTemplateBeta1({
+  Resources: {
+    Type: 'Foo::Bar',
+    Properties: {
+      Baz: 'Qux',
+    },
+  },
+});
+```
+
+#### Counting Resources
+
+This module allows asserting the number of resources of a specific type found
+in a template.
+
+```ts
+assert.assertResourceCountBeta1('Foo::Bar', 2);
+```
+
+#### Resource Matching
+
+Beyond resource counting, the module also allows asserting that a resource with
+specific properties are present.
+
+The following code asserts that the `Properties` section of a resource of type
+`Foo::Bar` contains the specified properties -
+
+```ts
+assert.assertResourceBeta1('Foo::Bar', {
+  Foo: 'Bar',
+  Baz: 5,
+  Qux: [ 'Waldo', 'Fred' ],
+});
+```
+
+The same method allows asserting the complete definition of the 'Resource'
+which can be used to verify things other sections like `DependsOn`, `Metadata`,
+`DeletionProperty`, etc.
+
+```ts
+assert.assertResourceBeta1('Foo::Bar', {
+  Properties: { Foo: 'Bar' },
+  DependsOn: [ 'Waldo', 'Fred' ],
+}, {
+  part: ResourcePartBeta1.COMPLETE,
+});
+```
+
+## FAQ
+
+### What are we launching today?
+
+Today, we are launching a new jsii module called `@aws-cdk/assertions`, that allows
+CDK users to run assertions on the synthesized CloudFormation template.
+
+This module is available to users in all CDK supported languages.
+
+### Why should I use this feature?
+
+You should use this feature if you would like to ensure that your CDK constructs
+and CDK applications render expected CloudFormation resources and properties.
+
+### How is this different from the existing `assert` module?
+
+The two modules provide the same set of features. However, the existing module
+was only available to javascript users.
+The new version launched today is available to users of all CDK supported
+languages available today and any that we will support in the future.
+
+### How do I get started?
+
+This module is available in the languages as per the table below -
+
+| Language              | Pkg Manager | Package                                       |
+| --------------------- | ------------|---------------------------------------------- |
+| typescript/javascript | npm         | `@aws-cdk/assertions`                         |
+| Java                  | Maven       | `software.amazon.awscdk.assertions`           |
+| Python                | PyPI        | `aws-cdk.assertions`                          |
+| .NET                  | NuGet       | `Amazon.CDK.Assertions`                       |
+| Go                    | -           | `github.com/aws/aws-cdk-go/awscdk/assertions` |
+
+### Is this feature available in the AWS CDK v2?
+
+This module is available in both the AWS CDK v1 and v2, just like all of our other
+modules.
+
+In CDKv2, this module will be available as part of `aws-cdk-lib`.
+The import statement you will need to use is -
+
+```ts
+import { assertions } from 'aws-cdk-lib';
```

Comment:
Really? We're adding a testing library to `aws-cdk-lib`? This seems like a mistake to me...

Is this decision final?

Reply:
This RFC is about making this decision.

Why is it a mistake to add a testing library to `aws-cdk-lib`?

Reply:
@skinny85 what's your arguments against making the assertion library part of the AWS CDK. It's a common practice to include assertion utilities in standard libraries (see node.js's 'assert' for example).

We want to standardize test assertions to be an intrinsic part of the AWS CDK so it can be used as a foundation for the ecosystem.

The _important_ side effect is that it including it in the AWS CDK itself will allow us to use it from within the AWS CDK codebase, even if/when we aws-cdk-lib becomes a truly monolithic codebase (after we stop supporting v1).

Reply:
Do we have any concerns around the size of the CDK package? If so, this could be an argument against including assertions in the main package, which would force users to deploy a testing library to a production environment, where it's not used. I know with Javascript we can use things like webpack, but I'm not sure whether this can be done in other languages. Then again, depending on the additional size, this may not even be relevant.

Reply:
We do have concerns around the size but the assertion library is negligible comparing to the two huge lambda layers we bundle into the library, which is where we should optimize (see [here](https://github.com/aws/aws-cdk/projects/10#card-54949219)).